cxx-qt-lib/src/qurl.rs

```diff
@@ -0,0 +1,175 @@
+/// look at a QUrl through a reference or smart pointer, as in `&QUrl`
+/// or `UniquePtr<QUrl>`.
+#[repr(C)]
+pub struct QUrl {
+    _pinned: PhantomData<PhantomPinned>,
+}
+
+impl QUrl {
+    /// Create a new Rust Url from this QUrl.
+    /// This is a copy operation so any changes will not propagate to the original QUrl.
+    pub fn to_rust(&self) -> Url {
+        Url::from_qurl(self)
+    }
+}
+
+// Safety:
+//
+// The code in this file ensures that QUrl can only ever be allocated
+// on the stack in pinned form which avoids the pitfalls of trying to
+// move this type that has a non-trivial move constructor.
+unsafe impl ExternType for QUrl {
+    type Id = type_id!("QUrl");
+    type Kind = cxx::kind::Opaque;
+}
+
+extern "C" {
+    #[link_name = "cxxqt1$unique_ptr$qurl$null"]
+    fn unique_ptr_qurl_null(this: *mut MaybeUninit<*mut c_void>);
+    #[link_name = "cxxqt1$unique_ptr$qurl$raw"]
+    fn unique_ptr_qurl_raw(this: *mut MaybeUninit<*mut c_void>, raw: *mut QUrl);
+    #[link_name = "cxxqt1$unique_ptr$qurl$get"]
+    fn unique_ptr_qurl_get(this: *const MaybeUninit<*mut c_void>) -> *const QUrl;
+    #[link_name = "cxxqt1$unique_ptr$qurl$release"]
+    fn unique_ptr_qurl_release(this: *mut MaybeUninit<*mut c_void>) -> *mut QUrl;
+    #[link_name = "cxxqt1$unique_ptr$qurl$drop"]
+    fn unique_ptr_qurl_drop(this: *mut MaybeUninit<*mut c_void>);
+}
+
+// Safety: TODO
+unsafe impl UniquePtrTarget for QUrl {
+    #[doc(hidden)]
+    fn __typename(f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("QUrl")
+    }
+
+    #[doc(hidden)]
+    fn __null() -> MaybeUninit<*mut c_void> {
+        let mut repr = MaybeUninit::uninit();
+        unsafe {
+            unique_ptr_qurl_null(&mut repr);
+        }
+        repr
+    }
+
+    #[doc(hidden)]
+    unsafe fn __raw(raw: *mut Self) -> MaybeUninit<*mut c_void> {
+        let mut repr = MaybeUninit::uninit();
+        unique_ptr_qurl_raw(&mut repr, raw);
+        repr
+    }
+
+    #[doc(hidden)]
+    unsafe fn __get(repr: MaybeUninit<*mut c_void>) -> *const Self {
+        unique_ptr_qurl_get(&repr)
+    }
+
+    #[doc(hidden)]
+    unsafe fn __release(mut repr: MaybeUninit<*mut c_void>) -> *mut Self {
+        unique_ptr_qurl_release(&mut repr)
+    }
+
+    #[doc(hidden)]
+    unsafe fn __drop(mut repr: MaybeUninit<*mut c_void>) {
+        unique_ptr_qurl_drop(&mut repr)
+    }
+}
+
+pub struct Url {
+    // Note that once map_qt_value is removed later, this can become private again
+    #[doc(hidden)]
+    pub inner: cxx::UniquePtr<QUrl>,
+}
+
+impl Url {
+    pub fn from_qurl(qurl: &QUrl) -> Self {
+        Self {
+            // Safety: TODO
+            inner: unsafe {
+                let mut ptr = MaybeUninit::<cxx::UniquePtr<QUrl>>::zeroed();
+                qurl_init_from_qurl(&mut ptr, qurl);
+                ptr.assume_init()
+            },
+        }
+    }
+
+    pub fn from_string(s: String) -> Self {
```

Comment:
Why does this take a `String`, instead of a `&str`?

Reply:
Probably rename to from_str then as well

Reply:
Done